File: src/storeFx/fetchAlbumInfo.ts
import * as Fx from '../fx/fx';
import { mergeIntoMap } from '../mergeIntoMap/mergeIntoMap';
import type { Store } from '../store/types';
import { getAlbumInfo2 } from '../subsonic';

export function fetchAlbumInfo(albumId: string) {
  return Fx.bind(Fx.ask<{ store: Store }>(), ({ store }) =>
    Fx.bind(
      Fx.provide(getAlbumInfo2(albumId), {
        credentials: store.getState().auth.credentials,
      }),
      albumInfo => {
        const state = store.getState();

        const infoById = mergeIntoMap(
          state.albums.infoById,
          [albumInfo],
          () => albumId,
        );

        if (infoById !== state.albums.infoById) {
          store.setState(prevState => ({
            albums: { ...prevState.albums, infoById },
          }));
        }

        return Fx.Ok();
      },
    ),
  );
}

File: src/storeFx/fetchAlbumInfo.ml
open Fx
open Monad_syntax
open MergeIntoMap
open Zustand

let make albumId =
  let* Deps.{ store } = ask () in
  let* albumInfo =
    let state = getState store in
    Subsonic.GetAlbumInfo2.make albumId
    |. provide ~deps:{ credentials = state.auth.credentials }
  in
  let infoById =
    let state = getState store in
    mergeIntoMap state.albums.infoById [| albumInfo |] (fun _ -> albumId)
  in
  if
    let state = getState store in
    infoById != state.albums.infoById
  then
    store
    |. Zustand.setState (fun prevState ->
        { prevState with albums = { prevState.albums with infoById } });
  Ok ()
